pumpkin-inventory/src/lib.rs

```diff
@@ -52,3 +62,185 @@ impl WindowType {
+
+    fn handle_item_change(
+        &mut self,
+        carried_item: &mut Option<ItemStack>,
+        slot: usize,
+        mouse_click: MouseClick,
+    ) -> Result<(), InventoryError> {
+        let mut all_slots = self.all_slots();
+        if slot > all_slots.len() {
+            Err(InventoryError::InvalidSlot)?
+        }
+        handle_item_change(carried_item, all_slots[slot], mouse_click);
+        Ok(())
+    }
+
+    fn all_slots(&mut self) -> Vec<&mut Option<ItemStack>>;
+
+    fn all_slots_ref(&self) -> Vec<Option<&ItemStack>>;
+
+    fn all_combinable_slots(&self) -> Vec<Option<&ItemStack>> {
+        self.all_slots_ref()
+    }
+
+    fn all_combinable_slots_mut(&mut self) -> Vec<&mut Option<ItemStack>> {
+        self.all_slots()
+    }
+
+    fn print_all_contents(&self) {
+        self.all_slots_ref()
+            .into_iter()
+            .enumerate()
+            .filter_map(|(slot, item)| item.map(|item| (slot, item)))
+            .for_each(|(slot, item)| {
+                dbg!(slot, item);
+            });
+    }
+
+    fn internal_pumpkin_id(&self) -> u64 {
+        0
+    }
+}
+
+pub fn handle_item_take(
+    carried_item: &mut Option<ItemStack>,
+    item_slot: &mut Option<ItemStack>,
+    mouse_click: MouseClick,
+) {
+    let Some(item) = item_slot else {
+        return;
+    };
+    let mut new_item = *item;
+
+    match mouse_click {
+        MouseClick::Left => {
+            *item_slot = None;
+        }
+        MouseClick::Right => {
+            let half = item.item_count / 2;
+            item.item_count -= half;
+            new_item.item_count = half;
+        }
+    }
+    *carried_item = Some(new_item);
+}
+pub fn handle_item_change(
+    carried_slot: &mut Option<ItemStack>,
+    current_slot: &mut Option<ItemStack>,
+    mouse_click: MouseClick,
+) {
+    match (current_slot.as_mut(), carried_slot.as_mut()) {
+        // Swap or combine current and carried
+        (Some(current), Some(carried)) => {
+            if current.item_id == carried.item_id {
+                combine_stacks(carried_slot, current, mouse_click);
+            } else if mouse_click == MouseClick::Left {
+                let carried = *carried;
+                *carried_slot = Some(current.to_owned());
+                *current_slot = Some(carried.to_owned());
+            }
+        }
+        // Put held stack into empty slot
+        (None, Some(carried)) => match mouse_click {
+            MouseClick::Left => {
+                *current_slot = Some(carried.to_owned());
+                *carried_slot = None;
+            }
+            MouseClick::Right => {
+                carried.item_count -= 1;
+                let mut new = *carried;
+                new.item_count = 1;
+                *current_slot = Some(new);
+            }
+        },
+        // Take stack into carried
+        (Some(_current), None) => handle_item_take(carried_slot, current_slot, mouse_click),
+        (None, None) => (),
+    }
+}
```

Comment:
I think replacing this as return value would be also better here as well

Reply:
I don't agree. This function ensures correct behaviour for anyone using it. Otherwise correct behaviour can not be enforced as easily.